src/submission/migration/migration-manager.ts

```diff
@@ -18,811 +18,804 @@
+			});
+
+			if (checkedSubmission) {
+				continue;
+			}
+
+			migration.checkedSubmissions.push({
+				programId: sub.programId,
+				id: sub._id,
+			});
+
+			const command: RevalidateClinicalSubmissionCommand = {
+				programId: sub.programId,
+				migrationId: migration._id as string,
+			};
+
+			const result = await submissionService.operations.revalidateClinicalSubmission(
+				command,
+				newSchema,
+				dryRun,
+			);
+
+			if (!result.submission) {
+				continue;
+			}
+
+			if (result.submission.state === 'INVALID_BY_MIGRATION') {
+				migration.invalidSubmissions.push({
+					programId: sub.programId,
+					id: sub._id,
+				});
+			}
+
+			migration = await migrationRepo.update(migration);
+		}
+		return migration;
+	};
+
+	// start iterating over paged donor documents records (that weren't checked before)
+	const checkDonorDocuments = async (
+		migration: DictionaryMigration,
+		newSchema: dictionaryEntities.SchemasDictionary,
+	) => {
+		let migrationDone = false;
+		if (!migration._id) {
+			throw new Error('Migration should have an id');
+		}
+		const migrationId = migration._id;
+		const dryRun = migration.dryRun;
+		const breakingChangesEntitesCache: { [versions: string]: ClinicalEntitySchemaNames[] } = {};
+		const coreFieldChangesEntitiesCache: { [versions: string]: ClinicalEntitySchemaNames[] } = {};
+
+		const missingEntityExceptionCache = createMissingEntityExceptionCache();
+
+		while (!migrationDone) {
+			let invalidCount = 0;
+			let validCount = 0;
+			const donors = await getNextUncheckedDonorDocumentsBatch(migrationId, 20);
+			const programsWithChanges: Set<string> = new Set(migration.programsWithDonorUpdates);
+
+			// no more unchecked donors ??
+			if (donors.length == 0) {
+				// mark migration as done
+				migrationDone = true;
+				break;
+			}
+
+			// check invalidation criteria against each one
+			for (const donor of donors) {
+				await updateCaches(
+					donor,
+					newSchema,
+					coreFieldChangesEntitiesCache,
+					breakingChangesEntitesCache,
+				);
+
+				const validationResult = await revalidateDonorClinicalEntities(
+					donor,
+					newSchema,
+					breakingChangesEntitesCache,
+				);
+
+				if (validationResult && validationResult.length > 0) {
+					// if invalid mark as invalid and update document metadata
+					if (!dryRun) {
+						const invalidDonor = await markDonorAsInvalid(donor, migrationId);
+						updateSetOfProgramsWithChanges(donor, invalidDonor, programsWithChanges);
+					} else {
+						await updateMigrationIdOnly(donor, migrationId);
+					}
+
+					migration.invalidDonorsErrors.push({
+						donorId: donor.donorId,
+						submitterDonorId: donor.submitterId,
+						programId: donor.programId,
+						errors: validationResult,
+					});
+
+					invalidCount += 1;
+					continue;
+				} else {
+					if (!dryRun) {
+						const updatedDonor = await updateSingleDonorCompletionStats(
+							donor,
+							missingEntityExceptionCache,
+						);
```

Comment:
Get update donor info including core completion stats, function uses cache for fetching exception data.